rfcs/20211118-support-implicit-concat/README.md

```diff
@@ -0,0 +1,139 @@
+on API level - it must be enabled on user side manually due to several reasons,
+primarily because of memory management.
+
+Let's consider them on an example in which two tensors are merged to create a
+bigger one:
+* It may happen that two inputs, being an output from other primitives, may have
+  incompatible memory layouts, e.g., `aBx32b` and `ABx16a32b`. In this case
+  reorder would be needed no matter what, thus, doesn't make sense to avoid
+  explicit concat which would execute such reorder as a part of operation. It
+  means that user oversees that formats are same or compatible for implicit
+  merging. Note that this holds for each architecture and even if input memory
+  formats are same for one architecture, it does not necessarily mean that the
+  other one would have exact format or match would happen.
+* But even if formats are same, it may happen that implicit concat is still
+  impossible, e.g., first input has padding. Despite in theory implicit concat
+  is still possible, in the reality it may bring only troubles since second
+  input would highly likely be unaligned in memory and may lead to various
+  "interesting" performance effects, or even crashes, so that using explicit
+  concat would be safer, and potentially offer better performance.
+
+Let's consider two scenarios that may happen when considering implicit concat:
+
+* Scenario 1, when concatenation should happen over the axis which is the
+  outermost dimension in memory. E.g., `aBcd16b` inputs and output with `n = 1`
+  and concatenating over channels `C`. In this simplified scenario, two inputs
+  are separate dense pieces, and primitives may be configured as usual in this
+  case. The only thing required is to provide a proper memory handle for a
+  second piece. In this case, everything can be enabled on user side as of
+  current oneDNN state, if using CPU raw pointers or GPU USM memory kind. For
+  OpenCL buffers this trick won't work, unfortunately.
+  ![Scenario 1](Scenario_1.jpg)
+* Scenario 2a, when concatenation should happen over the axis which is not the
+  outermost dimension in memory. This is a generic scenario. Same example but
+  for `n > 1` works here. Let's for simplicity take `n = 2`. In this case each
+  primitive which contributes to the final output would have non-trivial strides
+  between `n` points. oneDNN does not provide any API which would allow the user
+  to initialize a memory descriptor with both blocking descriptor and strides
+  representing not dense case. But `submemory` API may be used to create such
+  with two stages: initialize original memory descriptor with tags coming from
+  compute primitive or passed from previous layer, and then perform `submemory`
+  of this descriptor from the `concat` destination. This would achieve the
+  desired effect. This general scenario will utilize `offset0` and `strides` in
+  implementations internally but it requires support from the library and it's
```

Comment:
Can there be any problems with unaligned access to output memory?

Reply:
Generally, we test unaligned accesses in benchdnn since protection mode aligns on right side, not left. But some unknown side effects, probably, possible.